File: 2.Algorithm/77.Kakao/실패율(Kakao)/실패율.py
def solution(N, stages):
    counts = {i:0 for i in range(1,N+2)}
    for i in stages:
        counts[i] += 1
    del(counts[N+1])
    
    total = len(stages)
    for j in counts:
        total -= counts[j]
        if total:
            counts[j] = counts[j]/(total+counts[j])
        
    return sorted(counts, key=lambda x: -counts[x])


def solution(N, stages):
    counts = [0 for _ in range(N+1)]
    answer = []
    for i in stages:
        counts[i-1] += 1
    counts.pop()
    total = len(stages)
    for i in range(N):
        answer.append((i+1, counts[i]/total if total > 0 else 0))
        total -= counts[i]
    
    return list(map(lambda x: x[0], sorted(answer, key=lambda x: (-x[1], x[0]))))

File: 2.Algorithm/77.Kakao/실패율(Kakao)/실패율.java
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.stream.Collectors;

class Solution {
    public List<Integer> solution(int N, int[] stages) {
        
		HashMap<Integer, Double> counts = new HashMap<>();

		for (int i : stages) {
			if (counts.containsKey(i)) {
				counts.put(i, counts.get(i) + 1);
			} else
				counts.put(i, (double) 1);
		}


		int stage = stages.length;

		for (int i = 1; i <= N; i++) {
			if (counts.containsKey(i)) {
				stage -= counts.get(i);
				counts.put(i, (double) counts.get(i) / (double) (stage + counts.get(i)));
			} else {
				counts.put(i, (double) 0);
			}
		}
		counts.remove(N + 1);

		List<Double> d = counts.values().stream().collect(Collectors.toList());

		return counts.keySet().stream().sorted(Comparator.comparing(counts::get).reversed()).collect(Collectors.toList());

    }
}